src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs

```diff
@@ -3298,113 +3216,86 @@ internal virtual ImmutableArray<LabelSymbol> Labels
+                       typeArgumentsSyntax: default,
+                       typeArguments: default,
+                       invoked: true,
+                       indexed: false,
+                       bindingDiagnostics);
 
-                    if (candidateMethods.Any())
-                    {
-                        // we have found some valid methods, so stop searching
-                        break;
-                    }
-                    else
-                    {
-                        // the ones we found weren't viable. Try again, starting from the parent of where we found these
-                        patternType = patternType.GetNextBaseTypeNoUseSiteDiagnostics(null, Compilation, ref visitedTypes);
-                        lookupResult.Clear();
-                    }
-                }
-                else
+                if (boundAccess.Kind != BoundKind.MethodGroup)
                 {
-                    break;
+                    // the thing is not a method
+                    return PatternLookupResult.NotAMethod;
                 }
-            }
 
-            MethodSymbol patternMethod = PerformPatternOverloadResolution(patternType, candidateMethods, syntaxExpr, warningsOnly, diagnostics, syntaxTree, messageID);
-            candidateMethods.Free();
-            visitedTypes?.Free();
+                var analyzedArguments = AnalyzedArguments.GetInstance();
+                var patternMethodCall = BindMethodGroupInvocation(
+                    syntaxNode,
+                    syntaxNode,
+                    methodName,
+                    (BoundMethodGroup)boundAccess,
+                    analyzedArguments,
+                    bindingDiagnostics,
+                    queryClause: null,
+                    allowUnexpandedForm: false);
 
-            return patternMethod;
-        }
+                analyzedArguments.Free();
 
-        /// <summary>
-        /// Finds a pattern method implemented as an extension method, stopping at the first matching method
-        /// </summary>
-        /// <param name="typeExpr">The expression on which to search.</param>
-        /// <param name="methodName">Method to search for.</param>
-        /// <param name="syntaxExpr">The expression for which lookup is being performed</param>
-        /// <param name="diagnostics">Populated with binding diagnostics.</param>
-        /// <param name="messageID">The ID of the message to use when reporting diagnostics</param>
-        /// <returns>The desired method or null.</returns>
-        private MethodSymbol FindPatternExtensionMethod(BoundExpression typeExpr, string methodName, SyntaxNode syntaxExpr,
-                                                        DiagnosticBag diagnostics, MessageID messageID)
-        {
-            AnalyzedArguments arguments = AnalyzedArguments.GetInstance();
-            MethodSymbol patternMethod = null;
+                if (patternMethodCall.Kind != BoundKind.Call)
```

Comment:
Just for my edification--what other BoundKinds could be found here that would result in returning NotCallable?

Reply:
It could also be a DynamicInvocation if any of the arguments are dynamic. In the dispose case it can't be (we never pass params) but in the general case its something we have to consider.

I should add a test for an optional dynamic arg though, as that should work ok.